gum/backend-x86/gumstalker-x86.c

```diff
@@ -2030,6 +2092,46 @@ gum_exec_ctx_new (GumStalker * stalker,
 
   ctx->depth = 0;
 
+#ifdef HAVE_LINUX
+  ctx->excluded_calls = gum_metal_hash_table_new (NULL, NULL);
+
+  {
+    GumInterceptor * interceptor;
+    GumAttachReturn attach_ret;
+
+    interceptor = gum_interceptor_obtain ();
+    stalker->interceptor = interceptor;
+
+    /* This function may be statically linked and therefore not an export */
+    __gxx_personality_v0_impl = gum_find_function ("__gxx_personality_v0");
```

Comment:
This is only called once per thread stalked. So quite infrequently? Is the DebugSymbol performance that much of an overhead it may be significant if it's on a very cold code path? Should I change it to a `dlsym`?

Reply:
Hmm well I suppose we could to `gum_module_find_export_by_name (NULL, name)` and fall back to `gum_find_function (name)`. The reason I'm worried about this isn't per call, but per process lifetime, as it tends to be significant when stalking a short-lived process. (E.g. on i/macOS it might take a few seconds for the first call where things are being warmed up.)

Reply:
I had a problem when I tried this. The `_Unwind_Find_FDE` function although statically linked did have an export. However, the value returned by the DebugSymbol is different. The one returned by the export is never called, I guess maybe it's just a wrapper?

Reply:
Oh, you mean our own libunwind? If so that sounds like a mistake made by the integrator of the Frida SDK, where they don't limit the exported symbols to the bare minimum.

Reply:
I’d guess so. Unless we somehow end up with two copies of things?

Reply:
Well I don't think we should worry about that, we should expect the integrator (who's linking in frida-gum) to ensure they don't export any other symbols than strictly needed, e.g. like frida-server does with `-Wl,--version-script,frida-server.version` containing:

```
FRIDA_SERVER {
  local:
    *;
};
```

Reply:
I've put in the change as suggested, but whilst x64 is fine, I am still having a problem on x86. As we can see below,
both contain an implementation of `_Unwind_Find_FDE` of their own, and both depend on `libc`. However, on `x86` whilst `libc` exports `_Unwind_Find_FDE`, on `x64` it does not. Obviously when the tests make use of `_Unwind_Find_FDE` it uses it's in-built copy. But by preferring `gum_module_find_export_by_name` over `gum_find_function` to find the function, on `x86` it has hooked the wrong one.

```bash
$ readelf -sW ../build/tmp-linux-x86/frida-gum/tests/gum-tests | grep _Unwind_Find_FDE
   929: 000642f9     5 FUNC    LOCAL  DEFAULT   15 _Unwind_Find_FDE.cold
 73952: 0101e1f0   628 FUNC    LOCAL  DEFAULT   15 _Unwind_Find_FDE
$ ldd ../build/tmp-linux-x86/frida-gum/tests/gum-tests
        linux-gate.so.1 (0xf7fd0000)
        libdl.so.2 => /lib/i386-linux-gnu/libdl.so.2 (0xf68c5000)
        libresolv.so.2 => /lib/i386-linux-gnu/libresolv.so.2 (0xf68ac000)
        libpthread.so.0 => /lib/i386-linux-gnu/libpthread.so.0 (0xf6889000)
        libm.so.6 => /lib/i386-linux-gnu/libm.so.6 (0xf6784000)
        libc.so.6 => /lib/i386-linux-gnu/libc.so.6 (0xf6595000)
        /lib/ld-linux.so.2 (0xf7fd1000)
$ readelf -sW /lib/i386-linux-gnu/libc.so.6 | grep _Unwind_Find_FDE
  1052: 00146cc0   186 FUNC    GLOBAL DEFAULT   16 _Unwind_Find_FDE@@GCC_3.0
```

```bash
$ readelf -sW ../build/tmp-linux-x86_64/frida-gum/tests/gum-tests | grep _Unwind_Find_FDE
   665: 0000000000116c7f     5 FUNC    LOCAL  DEFAULT   15 _Unwind_Find_FDE.cold
 55018: 00000000010a32b0   595 FUNC    LOCAL  DEFAULT   15 _Unwind_Find_FDE
$ ldd ../build/tmp-linux-x86_64/frida-gum/tests/gum-tests
        linux-vdso.so.1 (0x00007ffff7fce000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007ffff67cb000)
        libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007ffff67af000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007ffff678c000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007ffff663d000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ffff644b000)
        /lib64/ld-linux-x86-64.so.2 (0x00007ffff7fcf000)
$ readelf -sW /lib/x86_64-linux-gnu/libc.so.6 | grep _Unwind_Find_FDE
$
```

Reply:
I have made my changes. You'll see I haven't squashed the commits this time to make it easier to check the changes. I don't think that on Linux at least I can find the statically linked `_Unwind_Find_FDE` symbol without using the `DebugSymbol` API. So I have reverted back to just providing an `extern` definition for the function in question and then letting the linker give me the address (this should provide me with the statically linked symbol if present, or an imported one if not).

You said this would cause issues on Android. I am not sure whether this same exception handling mechanism is used on Android? So for now I have guarded all the exception handling stuff (including the external function definitions) behind a:
```c
#if defined (HAVE_LINUX) && !defined (HAVE_ANDROID)
```.

The only other option I can think of, is if you think the mechanism should still work the same on Android, but that Android won't be affected by the same duplicate symbol issue, we could retain the exception handling mechanism but use compile time `#if` directives to have Linux use the `extern` function definitions, but have Android use a dynamic approach?

Let me know what you think. Cheers.